### PR TITLE
Add calculation of UUID for getEntry

### PR DESCRIPTION
**Pull request by loosebazooka**
- Calculate UUID by recontructing the entry
- Add option to getEntry by HashedRekorRequest(sig,cert,artifact) on RekorClient
  UUID is calculated client side (avoids a search on the rekor instance)

Signed-off-by: Appu Goundan <appu@google.com>

### CODE REVIEW

File: src/main/java/dev/sigstore/rekor/client/HashedRekordRequest.java [hunk 1]
```diff
@@ -57,20 +58,35 @@ public static HashedRekordRequest newHashedRekordRequest(
...
+    }
   }
 
   public HashedRekord getHashedRekord() {
-    return hashedrekord;
+    return hashedRekord;
+  }
+
+  /** Computes the expected rekor uuid of an entry based on the content of the hashedRekord. */
+  public String computeUUID() {
```
Comment (patflynn) [inline src/main/java/dev/sigstore/rekor/client/HashedRekordRequest.java:87]: It's a bit surprising that the UUID is actually just a hash.

File: src/main/java/dev/sigstore/rekor/client/RekorVerifier.java [hunk 2]
```diff
@@ -74,12 +72,21 @@ public void verifyEntry(RekorEntry entry)
...
+          Base64.getDecoder().decode(entry.getVerification().getSignedEntryTimestamp()))) {
+        throw new RekorVerificationException("Entry SET was not valid");
+      }
+    } catch (InvalidKeyException ike) {
+      throw new RekorVerificationException("Public Key was invalid", ike);
+    } catch (SignatureException se) {
+      throw new RekorVerificationException("Signature was invalid", se);
+    } catch (NoSuchAlgorithmException nsae) {
+      // we should never be here, the verifier algorithm is expected to exist
+      throw new RuntimeException(nsae);
```
Comment (patflynn) [inline src/main/java/dev/sigstore/rekor/client/RekorVerifier.java:89]: the comment could go as a message into the exception.  You could also use AssertionError to make it clearer this shouldn't happen.

Comment (loosebazooka) [reply src/main/java/dev/sigstore/rekor/client/RekorVerifier.java:89]: Word

File: src/main/java/dev/sigstore/rekor/client/HashedRekordRequest.java (hunk 1 above)
Comment (loosebazooka) [reply src/main/java/dev/sigstore/rekor/client/HashedRekordRequest.java:87]: Yeah it's how it works in trillian.

File: src/main/java/dev/sigstore/rekor/client/RekorVerifier.java (hunk 2 above)
Comment (loosebazooka) [reply src/main/java/dev/sigstore/rekor/client/RekorVerifier.java]: done

**Review by patflynn — DISMISSED** [review]

**Comment by loosebazooka** [comment]
ugh rebased to fix merge conflict.

File: src/main/java/dev/sigstore/rekor/client/RekorVerifier.java [hunk 3]
```diff
@@ -77,12 +75,22 @@ public void verifyEntry(RekorEntry entry)
...
-      throw new RekorVerificationException("Entry SET was not valid");
+    try {
+      var verifier = Signature.getInstance(verifierAlgorithm);
+      verifier.initVerify(rekorPublicKey);
+      verifier.update(signableJson.getBytes(StandardCharsets.UTF_8));
+      if (!verifier.verify(
+          Base64.getDecoder().decode(entry.getVerification().getSignedEntryTimestamp()))) {
+        throw new RekorVerificationException("Entry SET was not valid");
+      }
+      // TODO: Verify more than just "ec" signed rekor entries
```
Comment (loosebazooka) [inline src/main/java/dev/sigstore/rekor/client/RekorVerifier.java:86]: I messed something up here. Hang on.

**Review by patflynn — APPROVED** [review]

**Pull request merged**